### PR TITLE
[Win] Add support for VS2022 Windows build.

### PR DESCRIPTION
**Pull request by Blackhex**
The update of Windows bootstrapping scripts needed for https://github.com/pytorch/pytorch/pull/86591 and https://github.com/pytorch/pytorch/issues/87695.

### CODE REVIEW

**Comment by Blackhex** [comment]
This is still WIP, I just need an advice on how to run pipelines with the changes.

File: conda/build_pytorch.sh [hunk 1]
```diff
@@ -321,7 +321,7 @@ for py_ver in "${DESIRED_PYTHON[@]}"; do
 
     # We need to build the compiler activation scripts first on Windows
     if [[ "$OSTYPE" == "msys" ]]; then
-        vs_package="vs$VC_YEAR"
+        vs_package="vs$VS_YEAR"
```
Comment (Blackhex) [inline conda/build_pytorch.sh:324]: I am not sure about this. There is/was a discrepancy between `VS_YEAR` and `VC_YEAR` variable naming. From Conda standpoint, `VC_YEAR` does not exists vile `VS_YEAR` does (https://docs.conda.io/projects/conda-build/en/stable/user-guide/environment-variables.html).

File: conda/build_vision.sh [hunk 2]
```diff
@@ -83,7 +83,7 @@ else
 fi
 
 if [[ "$OSTYPE" == "msys" ]]; then
-    time conda build -c $ANACONDA_USER --no-anaconda-upload vs2017
+    time conda build -c $ANACONDA_USER --no-anaconda-upload vs2022
```
Comment (Blackhex) [inline conda/build_vision.sh:86]: This needs to be tested somehow?

File: windows/cuda116.bat [hunk 3]
```diff
@@ -16,7 +16,7 @@ IF ERRORLEVEL 1 goto :eof
 REM Check for optional components
 
 set USE_CUDA=
-set CMAKE_GENERATOR=Visual Studio 15 2017 Win64
+set CMAKE_GENERATOR=Visual Studio 17 2022 Win64
```
Comment (Blackhex) [inline windows/cuda116.bat:19]: I've updated only CUDA 11.6 and 11.7 as those are listed in the build selection table on the homepage (https://pytorch.org/)

File: windows/internal/vs2022_install.ps1 [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# https://developercommunity.visualstudio.com/t/install-specific-version-of-vs-component/1142479
+# https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-history#evergreen-bootstrappers
+
+# 17.3.6 BuildTools
+$VS_DOWNLOAD_LINK = "https://ossci-windows.s3.us-east-1.amazonaws.com/vs17.3.6_BuildTools.exe"
```
Comment (Blackhex) [inline windows/internal/vs2022_install.ps1:5]: The file currently does not exist, how to get it there?

**Review by malfet — CHANGES_REQUESTED** [review]
We should not add any new batch files as there are no way to abort script execution on error, which makes debugging any failures much harder. Please rewrite them either into PowerShell or better into Python

**Comment by Blackhex** [comment]
@malfet Sure, no problem, but please note that I currently don't have permissions to run PR pipelines on this repo.

**Pull request closed without merging**